Feature/prlist @qumei

zqumei0 opened the pull request:
*Issue #, if available:*
N/A
*Description of changes:*
Added new function prlist which lists all open pull requests for a specified repository. For more detail see updated README.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

zqumei-amzn committed:
Added prlist function

zqumei-amzn committed:
Fixed comment type

rkeene commented on git-coco:
```diff
@@ -164,6 +164,47 @@ pr() {
+        local dst_branch src_branch is_merge pull_id author title
+
+        IFS=$'\t'
+        read -r author dst_branch is_merge pull_id src_branch title <<< "${pull_info}"
```
These two lines should be merged so the `IFS` variable is only affecting the `read` command

zqumei0 replied:
Combined lines 197 and 198 to contain scope of IFS.

rkeene commented on git-coco:
```diff
@@ -164,6 +164,47 @@ pr() {
+
+    if [ -z "$repo" ]; then
+        fail "Usage: git coco prlist <repo>"
+    fi
```
What would be super cool is if this could check `git remote -v` and determine the repo if one isn't specified

There are 4 cases to cover:

1. `codecommit:` with no profile, just repo name
2. `codecommit:` with profile and repo name
3. `https:` that resolves to CodeCommit
4. Other

Example of 1:

```
origin	codecommit::us-east-1://Test1 (fetch)
origin	codecommit::us-east-1://Test1 (push)
```

Example of 2:

```
origin	codecommit::us-east-1://coco@Test2 (fetch)
origin	codecommit::us-east-1://coco@Test2 (push)
```

Example of 3:

```
origin	https://git-codecommit.us-east-1.amazonaws.com/v1/repos/Test3 (fetch)
origin	https://git-codecommit.us-east-1.amazonaws.com/v1/repos/Test3 (push)
```

Example of 4:

```
origin	https://github.com/rkeene/git-coco.git (fetch)
origin	https://github.com/rkeene/git-coco.git (push)
```

If the repo (and possibly profile) can be found from one of these patterns, then we could use that information to query AWS CodeCommit.

zqumei0 replied:
Added a url_parser() function to parse urls passed to it. Function sets _profile variable if a profile is found from a codecommit:// url. In all other cases it returns (via echo) the repository name parsed from the url.

prlist() calls url_parser() with the output of git remote -v passed in to be parsed.

Let me know if you think it performs what you've asked or if I missed something.

rkeene replied:
This looks cool, I added some more comments around it above.

Thanks !  :+1:

rkeene commented on git-coco:
```diff
@@ -164,6 +164,47 @@ pr() {
+        local pull_info=$(aws_cli codecommit get-pull-request \
+        --pull-request-id "$pull_request" \
+        --query "pullRequest.{PR: pullRequestId, Title: title, Merge: pullRequestTargets[0].mergeMetadata.isMerged, DstBranch: pullRequestTargets[0].destinationReference, SrcBranch: pullRequestTargets[0].sourceReference, Author: authorArn}" \
+        --output text)
```
This should probably be quoted

rkeene commented on git-coco:
```diff
@@ -164,6 +164,47 @@ pr() {
+        
+        author=$(echo $author | awk -F '/' '{print $NF}')
+        dst_branch=$(echo $dst_branch | awk -F 'refs/heads/' '{print $2}')
+        src_branch=$(echo $src_branch | awk -F 'refs/heads/' '{print $2}')
```
These should also probably be quoted.

It's also possible to do the replacements in pure bash here:

```sh
$ ( dst_branch='refs/heads/bug/test1'; echo "${dst_branch/refs\/heads\/}" )
bug/test1
```

zqumei0 replied:
Adjusted assignments of dst_branch and src_branch to use pure bash. Quoted all lines.

zqumei-amzn committed:
Added url_parser() function to parse repository name and profile based on codecommit:// or https:// url
Adjusted prlist to use git remote -v output if no repository name was specified instead of exiting out
Adjusted output comments to have [-] if output is due to some incorrect input

zqumei-amzn committed:
Quoted assignments of author, dst_branch, and src_branch variables

rkeene commented on git-coco:
```diff
@@ -11,6 +11,34 @@ fail() {
+        if [ -z "$repo" ]; then
+            repo=$profile
+        else
+            $_profile=$profile
```
The left-hand side shouldn't have a dollar sign (`$`) here and quoting is usually a good idea

```suggestion
            _profile="$profile"
```

rkeene commented on git-coco:
```diff
@@ -11,6 +11,34 @@ fail() {
+    fi
+    
+    echo $repo
+}
```
This function should probably return an empty string or something in the case where the repo isn't an AWS CodeCommit URL (case 4).

zqumei0 replied:
Added a second regex check for CodeCommit URLs to differentiate between case 3 and 4. Function returns empty repo variable if not a valid CodeCommit URL.

rkeene commented on git-coco:
```diff
@@ -11,6 +11,34 @@ fail() {
+    # https:// url
+    else
+        repo=$(echo $url | awk -F '/' '{print $NF}')
+        repo=$(echo "${repo/\.git}")
```
The `echo` is redundant here -- you can just assign to the variable replacement

```suggestion
        repo="${repo/\.git}"
```

rkeene commented on git-coco:
```diff
@@ -164,6 +192,50 @@ pr() {
+    if [ -z "$repo" ]; then
+        local url=$(echo $(git remote -v) | awk -F ' ' '{print $2}')
+        repo=$(url_parser "$url")
+        echo "[-] No repo provided. Using current directory"
```
To me, it would be better if it didn't print this out if it was successful in finding the repo from "`git remote`"

If `url_parser` can't find the AWS CodeCommit repository then it should probably print an error and exit.

rkeene commented on git-coco:
```diff
@@ -164,6 +192,50 @@ pr() {
+
+    # Handle default case
+    if [ -z "$repo" ]; then
+        local url=$(echo $(git remote -v) | awk -F ' ' '{print $2}')
```
This could be better written as:

```suggestion
        local url=$(git remote -v | awk -F ' ' '{ print $2; exit; }')
```

Which removes some of the coincidental feel of unquoted echo not expanding newlines.

Another option would be to loop through all the remotes until `url_parser` returned something.

zqumei0 replied:
Went for the loop implementation. Code now loops through each URL until a valid repository name is found. If not, it exits with an error.

rkeene commented on git-coco:
```diff
@@ -184,7 +256,10 @@ case "$command" in
+        ;;
     *)
-        fail "Usage: git coco ls|create|clone|rm|init|pr"
+        fail "[-] Usage: git coco ls|create|clone|rm|init|pr|prlist"
```
It's just my opinion, but I don't like the `[-]` part because it's not a standard part of other tools -- when I see it I have to figure out why it's there when no other tool has it.

rkeene commented on git-coco:
```diff
@@ -11,6 +11,34 @@ fail() {
+
+    # https:// url
+    else
+        repo=$(echo $url | awk -F '/' '{print $NF}')
```
I would quote the URL here (as well as verifying that it belongs to AWS CodeCommit, as noted below)

zqumei-amzn committed:
Resolved PR comments

zqumei-amzn committed:
Added information to README.md on prlist command

rkeene commented on git-coco:
```diff
@@ -164,6 +194,61 @@ pr() {
+
+    local header="%-10s %-60s %-8s %-100s\n"
+    local format="%-10s %-60s %-8s %s <- %s (%s)\n"
+    local pr_query="pullRequest.{PR: pullRequestId, Title: title, Merge: pullRequestTargets[0].mergeMetadata.isMerged, DstBranch: pullRequestTargets[0].destinationReference, SrcBranch: pullRequestTargets[0].sourceReference, Author: authorArn}"
```
I think the "isMerged" field isn't useful here since the PR is open (`--pull-request-status OPEN` is specified below) -- what would be useful (but require additional API calls) is if it's mergable, but maybe it's better to just reduce the information shown

zqumei0 replied:
Good point, I'll leave it as is with just pulling out merge status. If anything pops up that seems useful to include in that table I'll probably add it in the future.

zqumei-amzn committed:
Removed merge status relevant code.

zqumei0 commented:
Rebased on #6
